Determine resolution from input

mikinakajima opened the pull request:
I am sending a pull request, that determines the resolution in the input file.
PLEASE DO NOT MERGE YET. We should probably merge Tyler's feature (determine the directory of the output file) first because I am having a problem with generating an output directory name.

mikinakajima committed:
GI.ph Update

mikinakajima committed:
Update GI.h

mikinakajima committed:
Updated GI.h

mikinakajima committed:
Update GI.h

mikinakajima committed:
Updated GI.h

mikinakajima committed:
Updated GI.h

mikinakajima committed:
Updated GI.h

mikinakajima committed:
Updated GI.h

mikinakajima committed:
Updated input file

mikinakajima committed:
Removing unnecessary files

mikinakajima commented:
I have a quick question. I sent a pull request, but found some files that should be removed. I removed these files, committed,  and pushed. However, I can still see the removed files in the "Files Exchanged" Section on the website. Is there anything I could do the update the "Files Exchanged"?

gassmoeller commented:
If I look at Files changed I only see `GI.modified.h`. The other files (GI.integrated.h and so on) that you removed in your last commit are not longer there. Did you just forget to remove GI.modified.h?

mikinakajima committed:
Delete GI.modified.h

mikinakajima commented:
You are right. Sorry! I forgot to remove GI.modified.h.

mikinakajima committed:
Merge branch 'master' into Determine_Resolution_from_Input

mikinakajima committed:
Delete run.sh

mikinakajima committed:
Merge branch 'Determine_Resolution_from_Input' of https://github.com/NatsukiHosono/FDPS_SPH into Determine_Resolution_from_Input

mikinakajima commented:
I updated the code so that there is no conflict with the master version. Would you mind taking a look at this pull request?

gassmoeller commented:
I will do that. It is quite a bit of code though and I have to postpone it to tomorrow (talk deadline coming up). Will be on it tomorrow afternoon.

mikinakajima commented:
Sure, thank you so much :) !!

gassmoeller commented on input/input.txt:
```diff
@@ -0,0 +1,11 @@
+mode = 2
```
We should talk about where to store these input files. Currently you have both the `examples` and the `input` folder. I would prefer having them all in one place, and now that we have the possibility to choose the input folder we are not limited to where it is.

mikinakajima replied:
good point. I was sort of ignoring examples folder. Perhaps keeping everything in examples may be a good idea! But I will follow the convention.

gassmoeller commented on src/GI.h:
```diff
@@ -21,15 +21,19 @@ template <class Ptcl> class GI : public Problem<Ptcl>{
-        end_time = parameter_file.getValueOf("end_time",1.0e+4);
-        damping = parameter_file.getValueOf("damping",1.);
+
+		int mode = parameter_file.getValueOf("mode", 2 );
```
while you are here, please make this `const unsigned int mode = ...`.

gassmoeller commented on src/GI.h:
```diff
@@ -41,8 +45,14 @@ template <class Ptcl> class GI : public Problem<Ptcl>{
 		const double offset = 5.0 * UnitRadi;
-		const PS::F64 dx = 1.0 / 39;
+		// the following line predicts the number of grid points in one direction
+		const int  gridpoint = int(2.0/pow(4.18*1.1/Nptcl,0.333));
```
These are a lot of "magic numbers". It would be easier to understand with the following:
```suggestion
                // Put a comment here what this formula is, in particular: why is it 2 / (V^1/3)? And why is there a factor of 1.1 in there? 
		const unsigned int n_gridpoints = static_cast<unsigned int>(2.0/pow(4.0/3.0*math::pi*1.1/Nptcl,1.0/3.0));
```

mikinakajima replied:
Thank you - I added some comments on this

gassmoeller replied:
Thanks for the documentation. Could you still change `4.18` to `4.0/3.0 * math::pi`? I know it is just a guess, but no need to obscure the formula behind the guess.

gassmoeller commented on src/GI.h:
```diff
@@ -41,8 +45,14 @@ template <class Ptcl> class GI : public Problem<Ptcl>{
+		// the following line predicts the number of grid points in one direction
+		const int  gridpoint = int(2.0/pow(4.18*1.1/Nptcl,0.333));
+		const PS::F64 dx =  2.0/gridpoint;
+		//const PS::F64 dx = 1.0 / 39;
```
Remove this line.

gassmoeller commented on src/GI.h:
```diff
@@ -106,7 +116,7 @@ template <class Ptcl> class GI : public Problem<Ptcl>{
 		const int tarNptcl = tarNcore + tarNmntl;
 		const int impNptcl = impNcore + impNmntl;
-		const int Nptcl    = tarNptcl + impNptcl;
+		//const int Nptcl    = tarNptcl + impNptcl;
```
Remove this line

gassmoeller commented on src/GI.h:
```diff
@@ -106,7 +116,7 @@ template <class Ptcl> class GI : public Problem<Ptcl>{
 		///////////////////
 		const int tarNptcl = tarNcore + tarNmntl;
```
It seems to me that with your changes this number, and the output that happens in the lines below are not longer correct. Is there a way to fix this? Maybe put the screen output below the place where the particles are created, and use the real number of core and mantle particles (not the one created in the simulation above)?

mikinakajima replied:
Thanks -- now the code generates these numbers after generating a planet

gassmoeller commented on src/GI.h:
```diff
@@ -166,15 +177,63 @@ template <class Ptcl> class GI : public Problem<Ptcl>{
-                for(PS::U32 i = 0 ; i < imp.size() ; ++ i){
-                    imp[i].mass /= (PS::F64)(Nptcl);
-                }
+		// I do not think we need this
```
Can you explain why? Where is the mass of each particle set instead? Or does the absolute mass not matter, and only the relative mass between particles matters?

TylerLaBree replied:
Miki and I suspect that this was a bug I implemented some time ago. This code currently takes the mass of a particle set by an output file, which should be correct, and then divides it by the total body mass. I think this step should only be done once, at the first creation of the particle.

gassmoeller replied:
Oh, I did not realize this is the part of the code that reads in particle data from a file. Then you are right, get rid of it.

gassmoeller commented on src/GI.h:
```diff
@@ -166,15 +177,63 @@ template <class Ptcl> class GI : public Problem<Ptcl>{
+	      {
                 //Put Tar.
+
+	      if (tarNmntl <  int(Nptcl * (1.0-coreFracMass))){
```
I do not understand this if-condition. `tarNmntl` is the number of particles generated in the mantle if there would be no particles removed. `Nptcl` is the total number of particles to be generated. How does this condition check if there are too few particles?
As a technical side note: `int()` is not a good way to convert types in C++. It can fail with weird error if you put in types that can not be converted to an integer, for example objects. A better alternative is `static_cast<int>(Nptcl * ...)`.

gassmoeller commented on src/GI.h:
```diff
@@ -166,15 +177,63 @@ template <class Ptcl> class GI : public Problem<Ptcl>{
+	      }
+
+
+	      if (tarNcore <  int(Nptcl *coreFracMass)){
```
Same here, can you explain in a comment how this if works?

mikinakajima replied:
Yes - at times you may produce tarNmntl which is smaller than Nptcl (your ideal particle number). In that case, you will have to decrease the grid spacing (dx) (or increase the resolution). I will make it clearer in the next version

gassmoeller commented on src/GI.h:
```diff
@@ -166,15 +177,63 @@ template <class Ptcl> class GI : public Problem<Ptcl>{
+	      }
+
+
+	      //removing particles to reach the exact Nptcl
```
This looks like a two-stage removal process two me. First you remove particles in regular intervals. Then if you need to remove any more you remove random numbers until you have as many selected as you want to remove. Is that what happens? Since you do this twice (once for the mantle, once for the core), I would recommend moving this into a separate function. Here is how that could look like (I have not tested this, let me know if something does not work):
```
std::unordered_set
create_removal_list (const unsigned int highest_index,
                                 const unsigned int number_of_removed_items){

              std::unordered_set<unsigned int> removal_list;

             if (number_of_removed_items == 0)
               return removal_list;

              // First remove items in regular intervals from the list
	      const unsigned int interval = int(highest_index / number_of_removed_items);
	      for (unsigned int i=0; i < highest_index; i += interval){
		removal_list.insert(i);
	      }

	      // If the removal list is shorter than the target number
              // (i.e. the highest index is not divisable by the number of
              // removed items), remove random items until the size fits
	      while (removal_list.size() < number_of_removed_items){
		const unsigned int num = rand ()  % static_cast<unsigned int>(highest_index);
                // This works even if num is already in the list, because the unordered_set filters out duplicates
		removal_list.insert(num);
             }
}
```

For the unordered set you will have to include the following at the top of the file (see my comment below on what an unordered set is): `#include <unordered_set>`

mikinakajima replied:
Yes, you are right -- this is a two step process. I thought generating random numbers for all the list may be slow, so I did this way. But if that's not the case let me know!

gassmoeller commented on src/GI.h:
```diff
@@ -166,15 +177,63 @@ template <class Ptcl> class GI : public Problem<Ptcl>{
+	      // if the removal list is shorter than the target number
+	      while (removal_list.size() < tarNmntl -  int(Nptcl * (1.0-coreFracMass))){
+		num = rand ()  % int(tarNmntl) + 1;
+		if (std::count(removal_list.begin(),removal_list.end(),num) == 0) removal_list.push_back(num);
```
This is an extremely slow line, because for every new number you have to go through the whole list of already marked numbers. If you want to keep this random algorithm we should talk about using an `std::unordered_set` instead of an `std::vector`. Here is a description of what a std::unordered_set is: http://www.cplusplus.com/reference/unordered_set/unordered_set/. In essence it is like a vector, but it is faster to insert or remove arbitrary values in a big set.

mikinakajima replied:
This is powerful. Thank you! I will use this

gassmoeller commented on src/GI.h:
```diff
@@ -166,15 +177,63 @@ template <class Ptcl> class GI : public Problem<Ptcl>{
+		if (std::count(removal_list.begin(),removal_list.end(),num) == 0) removal_list.push_back(num);
+	      }
+
+	      std::sort(removal_list.begin(), removal_list.end());
```
Why do you need this sort?

gassmoeller commented on src/GI.h:
```diff
@@ -192,10 +251,50 @@ template <class Ptcl> class GI : public Problem<Ptcl>{
                             ith.tag = 0;
-                            if(ith.id / NptclIn1Node == PS::Comm::getRank()) tar.push_back(ith);
+
+			    if ((index - removal_list[index_removal]==0) && (index_removal < removal_list.size())){			     
```
I feel like I do not understand this routine, could you add a few comments? I would have expected something like this:
```
// See if the index is in the removal list, if so do not create particle
if (std::count(removal_list.begin(),removal_list.end(),ith.id) != 0){
  --id;
  continue;
}
else{
if(ith.id / NptclIn1Node == PS::Comm::getRank()) tar.push_back(ith);
}

gassmoeller replied:
If you change this to an unordered set the if condition would be even simpler:
```
if (removal_list.count(ith.id) == 1){
 ...
else
  ...
```

gassmoeller commented on src/GI.h:
```diff
@@ -192,10 +251,50 @@ template <class Ptcl> class GI : public Problem<Ptcl>{
+
+		// making the core condition
+		// initialize the removal list
+		removal_list.erase(removal_list.begin(),removal_list.end());
```
This is not wrong, but `removal_list.clear();` is simpler.

gassmoeller reviewed (commented):
Thanks, this looks it does what you want to achieve. I have some structural comments, and I currently do not quite understand the algorithm in a few parts. Please add some comments at the places I indicated. The suggestion about moving the algorithm to a separate function is only a suggestion, but I think it would make the code cleaner. And we need to fix the output before including this, I would not want to write a wrong number of particles to the screen.
Also, I think you accidentally committed a test_output.txt file in src/init/test_output.txt, please remove that.

Let me know if my comments are confusing (there is some higher-level c++ involved to make this work nicely) or if you want me to take another look.

mikinakajima commented:
Thank you so much Rene! I will look into the details before the meeting today. I will fix them today and tomorrow.

mikinakajima committed:
Delete test_output.txt

mikinakajima committed:
Updated GI.h

mikinakajima committed:
Merge branch 'master' into Determine_resolution_from_input

mikinakajima committed:
Updated GI.h

mikinakajima committed:
remove test directory

mikinakajima committed:
Delete input.txt

mikinakajima commented:
I think I incorporated all the changes -- let me know what you think. unordered_set is awesome!

Also - is it okay to commit very frequently? Does it make it a bit annoying to track changes?? Or, is it better to make a large set of changes and commit just once?

gassmoeller commented on src/GI.h:
```diff
@@ -1,10 +1,31 @@
+
+      // If the removal list is shorter than the target number
+      // (i.e. the highest index is not divisable by the number of
+      // removed items), remove random items until the size fits
```
I think with the new version of the function you can remove everything in the parenthesis.

gassmoeller commented on src/GI.h:
```diff
@@ -21,15 +42,19 @@ template <class Ptcl> class GI : public Problem<Ptcl>{
+		PS::F64 impVel = parameter_file.getValueOf("impVel",0.);
+		end_time = parameter_file.getValueOf("end_time",1.0e+4);
+		damping = parameter_file.getValueOf("damping",1.);
+		PS::F64 Nptcl  = parameter_file.getValueOf("Nptcl", 100000);
```
Could you give this parameter a more descriptive name? `total_number_of_particles` maybe? You can keep the internal variable name, but since we do not have a way to attach documentation to the parameters in the input file they should be as self-explanatory as possible.

gassmoeller commented on src/GI.h:
```diff
@@ -106,26 +141,6 @@ template <class Ptcl> class GI : public Problem<Ptcl>{
 		///////////////////
 		const int tarNptcl = tarNcore + tarNmntl;
 		const int impNptcl = impNcore + impNmntl;
```
What about this line and the one above? Would they also need to move down? At least they would need to be adjusted after creating the particles, because otherwise we still output the wrong numbers, we just output them later.

gassmoeller commented on src/GI.h:
```diff
@@ -166,15 +182,37 @@ template <class Ptcl> class GI : public Problem<Ptcl>{
+	      {
+		// checking if there are enough mantle particles
+	      if (tarNmntl <  static_cast<int>(Nptcl * (1.0-coreFracMass))){
+		std::cout << "Too few mantle particles. Increase the grid size. The easiet fix is to increase the gridpoint in GI.h " << std::endl;
```
```suggestion
		std::cout << "Too few mantle particles. Increase the grid size. The easiest fix is to increase the gridpoint in GI.h " << std::endl;
```

gassmoeller commented on src/GI.h:
```diff
@@ -166,15 +182,37 @@ template <class Ptcl> class GI : public Problem<Ptcl>{
+		exit(0);
+	      }
+
+	      // cheking if there are enough core particles
```
```suggestion
	      // checking if there are enough core particles
```

gassmoeller commented on src/GI.h:
```diff
@@ -166,15 +182,37 @@ template <class Ptcl> class GI : public Problem<Ptcl>{
+	      }
+
+	      // cheking if there are enough core particles
+	      if (tarNcore <  static_cast<int>(Nptcl *coreFracMass)){
```
```suggestion
	      if (tarNcore <  static_cast<int>(Nptcl * coreFracMass)){
```

gassmoeller commented on src/GI.h:
```diff
@@ -192,10 +230,27 @@ template <class Ptcl> class GI : public Problem<Ptcl>{
                     }
                 }
+
+		std::cout << "# of mantle particles = " <<  tar.size() << std::endl;
```
I think this is not true in MPI parallel runs. If you have multiple MPI processes, `tar` on each one of them will only have a fraction of the total number of particles. I think this should be:
```suggestion
		std::cout << "# of mantle particles = " << id << std::endl;
		const unsignd int targetMantleParticles = id;
```

gassmoeller commented on src/GI.h:
```diff
@@ -212,18 +267,30 @@ template <class Ptcl> class GI : public Problem<Ptcl>{
                     }
                 }
+
+		std::cout << "# of total particles = " <<  tar.size() << std::endl;
```
This should probably also be `id`

gassmoeller commented on src/GI.h:
```diff
@@ -192,10 +230,27 @@ template <class Ptcl> class GI : public Problem<Ptcl>{
+			    }else{
+			      if(ith.id / NptclIn1Node == PS::Comm::getRank()) tar.push_back(ith);
+			    }
+			    index += 1;
```
What is the difference between `id` and `index`?

mikinakajima replied:
id is counting the number of the target particles (up to Nptcl * (1.0-coreFracMass)), and index is counting the total number of the loop (up to tarNmntl)

gassmoeller commented on src/GI.h:
```diff
@@ -273,8 +340,34 @@ template <class Ptcl> class GI : public Problem<Ptcl>{
+		  std::cout << "Tar-to-Imp mass ratio: " << (double)(impNmntl) / (double)(tarNmntl) << std::endl;
+		}
+		
+		std::cout << "Total:" << Nptcl << std::endl;
```
Please add the following:
```suggestion
		assert(Nptcl == tarNptcl + impNptcl);
		std::cout << "Total:" << Nptcl << std::endl;
```
I am a bit worried that Nptcl and tarNptcl + impNptcl can become inconsistent, because we determine them in different ways. You may need to include `#include <assert.h>` at the top of the file for this.

gassmoeller commented on src/GI.h:
```diff
@@ -304,4 +397,10 @@ template <class Ptcl> class GI : public Problem<Ptcl>{
+
+
+
+	
```
Please remove these empty lines.

gassmoeller reviewed (commented):
Yes this is much better. I have a few remaining questions, mostly small things, but I am still worried that the screen output might be inconsistent in some cases. Please take a look at my comments.

gassmoeller commented:
Oh, and also: You can make as many commits as you like. The usual rule-of-thumb is to make commits as small as possible, but not smaller than a working logical unit. I.e. every commit should compile and contain a small change. I think your commit frequency is very reasonable (you could do better with the messages though, `update ...` is not a helpful message :-)).

mikinakajima committed:
Updated an algorithm to generate a target (pull request #33)

mikinakajima committed:
Merge branch 'Determine_Resolution_from_Input' of https://github.com/NatsukiHosono/FDPS_SPH into Determine_Resolution_from_Input

mikinakajima committed:
modified GI.h such that it can calculate tarNmntl, tarNcore, impNmntl, impNcore from tar.dat and imp.dat

This works for mode = 2, but it experiences a segmentation fault after a few time steps for mode = 1. I am not sure why

mikinakajima commented:
Thank you so much Rene! I think I incorporated the suggestions. 
I made addition changes; I moved the dummy algorithm only to mode 2, because mode 1 does not use this. Also, I suggested that we remove mode 3 because we are not using it (we can use mode 2 to generate target and impactor with separate runs). 

This update seems to be working for mode = 2, but not for mode = 1 ... after a few time steps, the code experiences a segmentation fault. I am not sure why.... I will look into it more later. Please do not merge yet.

mikinakajima committed:
Removed a big in reading the target mass

mikinakajima commented:
I may have fixed the issue, but I have to double check.

Also, for mode = 1, the imapctor location has to be moved to actually simulate an impact. I will do so (but I may not have time until Monday :( we shall see)

mikinakajima committed:
fixed mode =1 so that an impact can be simulated

mikinakajima commented:
Hi - would someone mind checking this branch and see if this can be merged with the master branch? My graduate student Scott Hull just arrived (yay!) and we would like to make progress on this issue, so that we can submit it this fall/winter. Thank you!

gassmoeller commented on src/GI.h:
```diff
@@ -1,10 +1,29 @@
+      
+      while (removal_list.size() < number_of_removed_items){
+	const unsigned int num = rand ()  % static_cast<unsigned int>(highest_index - lowest_index) + lowest_index;
+	//std::cout << num << std::endl;
```
Please remove the commented line.

mikinakajima replied:
did it - thanks!

gassmoeller commented on src/GI.h:
```diff
@@ -41,99 +67,40 @@ template <class Ptcl> class GI : public Problem<Ptcl>{
+		  Nptcl = (volume of a sphere)/(dx)^3 * L^3, where L = 2.0
+	          dx = (4.0/3.0 * math::pi/Nptcl)^{1/3}
+		  The number of grid point is 2.0/dx
+		  we multiply 1.1 so that the enough particles are created to generate a sphere */
```
```suggestion
		  we multiply by 1.1 so that enough particles are created to generate a sphere */
```

gassmoeller commented on src/GI.h:
```diff
@@ -163,18 +138,92 @@ template <class Ptcl> class GI : public Problem<Ptcl>{
                     ith.readAscii(impFile);
-                    ith.vel.x += (-1) * impVel;
+
+		    // This needs to be updated -- the code should determine the impactor and target sizes
```
Is this comment still up to date? The code does seem to take into account `impRadi` and `tarRadi`below. Or do you mean something else with the comment? If so please expand the comment.

gassmoeller commented on src/GI.h:
```diff
@@ -163,18 +138,92 @@ template <class Ptcl> class GI : public Problem<Ptcl>{
                 }
+
+		for(PS::U32 i = 0 ; i < imp.size() ; ++ i){
+		  if (imp[i].tag==0){
```
Could you not move this into the loop right above in line 155?

gassmoeller commented on src/GI.h:
```diff
@@ -189,13 +238,28 @@ template <class Ptcl> class GI : public Problem<Ptcl>{
+			    if (removal_list.count(index)){
+			      id += -1;
+			    }else{
+			      if(ith.id / NptclIn1Node == PS::Comm::getRank()) tar.push_back(ith);
```
Please beak this into two lines for readability.

mikinakajima replied:
I think I solved this, but let me know if I did not do it right..

gassmoeller commented on src/GI.h:
```diff
@@ -303,5 +363,5 @@ template <class Ptcl> class GI : public Problem<Ptcl>{
 			sph_system[i].acc += - sph_system[i].vel * 0.05 / sph_system[i].dt;
 		}
-	}
+	}	
```
Please remove the tab that you added at the end of the line.

gassmoeller reviewed (commented):
I am sorry it took so long to get a review on this. For the future: When you push a commit to a pull request, I do not get a message about it, I only get a message when you write a comment. In other words I was not aware you fixed the issue after your last comment.
I have a few minor comments, and only one remaining question:
- What is the change to examples/input.txt? Is this a new and more interesting setup? Was the old example useful? If yes to both, consider making it a new file or if you want to keep the name a new folder, e.g. `collision.txt`, or `examples/collision/input.txt`.

mikinakajima committed:
clean up GI.h

mikinakajima committed:
update Gi.h

mikinakajima commented:
I changed the input file because we had to write the whole input parameters otherwise the code produces an error. This need to be resolved in a future version.

mikinakajima commented:
Let me know if this is ready to be merged. :)

gassmoeller commented:
Thanks for all the effort! Looks very good now. :+1:
